stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,637 @@
+    if (rootElement == null) {
+      // null for rootElement is not allowed. We could support it, but our current
+      // implementation won't ever run into this, so let's punt on it for now.
+      throw new IllegalStateException();
+    }
+
+    if (mShadowDocument == null) {
+      mShadowDocument = new ShadowDocument(mDocumentProvider.getRootElement());
+      createShadowDOMUpdate().commit();
+      mDocumentProvider.setListener(new ProviderListener());
+    } else if (rootElement != mShadowDocument.getRootElement()) {
+      // We don't support changing the root element. This is handled differently by the
+      // protocol than updates to an existing DOM, and we don't have any case in our
+      // current implementation that causes this to happen, so let's punt on it for now.
+      throw new IllegalStateException();
+    }
+
+    return rootElement;
+  }
+
+  public void findMatchingElements(String query, Accumulator<Integer> matchedIds) {
+    verifyThreadAccess();
+
+    final Pattern queryPattern = Pattern.compile(Pattern.quote(query), Pattern.CASE_INSENSITIVE);
+    final Object rootElement = mDocumentProvider.getRootElement();
+
+    findMatches(rootElement, queryPattern, matchedIds);
+  }
+
+  private void findMatches(Object element, Pattern queryPattern, Accumulator<Integer> matchedIds) {
+    final ElementInfo info = mShadowDocument.getElementInfo(element);
+
+    for (int i = 0, size = info.children.size(); i < size; i++) {
+      final Object childElement = info.children.get(i);
+
+      if (doesElementMatch(childElement, queryPattern)) {
+        matchedIds.store(mObjectIdMapper.getIdForObject(childElement));
+      }
+
+      findMatches(childElement, queryPattern, matchedIds);
+    }
+  }
+
+  private boolean doesElementMatch(Object element, Pattern queryPattern) {
+    AttributeListAccumulator accumulator = acquireCachedAttributeAccumulator();
+    NodeDescriptor descriptor = mDocumentProvider.getNodeDescriptor(element);
+
+    descriptor.getAttributes(element, accumulator);
+
+    for (int i = 0, N = accumulator.size(); i < N; i++) {
+      if (queryPattern.matcher(accumulator.get(i)).find()) {
+        releaseCachedAttributeAccumulator(accumulator);
+        return true;
+      }
+    }
+
+    releaseCachedAttributeAccumulator(accumulator);
+    return queryPattern.matcher(descriptor.getNodeName(element)).find();
+  }
+
+  private ChildEventingList acquireChildEventingList(
+      Object parentElement,
+      DocumentView documentView) {
+    ChildEventingList childEventingList = mCachedChildEventingList;
+
+    if (childEventingList == null) {
+      childEventingList = new ChildEventingList();
+    }
+
+    mCachedChildEventingList = null;
+
+    childEventingList.acquire(parentElement, documentView);
+    return childEventingList;
+  }
+
+  private AttributeListAccumulator acquireCachedAttributeAccumulator() {
+    if (mCachedAttributeAccumulator == null) {
```

Comment:
This method has a bug. We're using a singleton caching system here, but this implementation will hand out the same `AttributeListAccumulator` more than once if its called twice without a call to `releaseCachedAttributeAccumulator` in-between.